### PR TITLE
Update custom collection documentation (leverage new collection protocol)

### PR DESCRIPTION
**Pull request by douglasdavis**
Follow up to #8674 - this PR updates the custom collection docs leveraging the inline documentation in the new `dask.typing.DaskCollection` protocol. I've cut out the "by hand" documentation of the collection methods (and in the process improved some of the protocol class docstrings). I also updated the example `Tuple` collection (now reflects the builtin collections a bit more, with a `__dask_optimize__` defined outside of the class).

- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: docs/source/custom-collections.rst [hunk 1]
```diff
@@ -563,6 +448,12 @@ Demonstrating this class:
     >>> x2.compute()
     (2, 3, 4, 5)
 
+    # Run-time typechecking from Python 3.8's Protocol and
+    # typing.runtime_checkable decorator.
```
Comment (jcrist) [inline docs/source/custom-collections.rst:452]: The user doesn't need to know anything about `runtime_checkable` or protocols, just that they can check if something is a dask collection using `isinstance(x, DaskCollection)`. I think this comment could be deleted, or updated, up to you.

**Review by jcrist — COMMENTED** [review]
Assuming the docs build and render properly, this seems mostly fine to me. Just one small nit.

File: docs/source/custom-collections.rst (hunk 1 above)
Comment (douglasdavis) [reply docs/source/custom-collections.rst:452]: Yes that makes sense, I'll trim that out

**Review by jcrist — APPROVED** [review]

**Comment by douglasdavis** [comment]
Thanks, Jim!

**Pull request merged**